File: assign1/go/perfect_number.go
package perfect_number

import(
  fp "github.com/repeale/fp-go"
)

func IsPerfectNumberImperative(number int) bool {
  sumOfFactors := 0

  for i := 1; i < number + 1; i++ {
    if number % i == 0 {
      sumOfFactors += i
    }
  }

  return number > 1 && sumOfFactors == number * 2
}

func IsPerfectNumberFunctional(number int) bool {
  all_numbers_in_range := fp.FlatMapWithIndex(func(_ int, i int) []int { return []int{i + 1}})(make([]int, number))

  factors := fp.Filter(func(x int) bool {return number % x == 0})(all_numbers_in_range)

  return number > 1 && fp.Reduce(func(sum int, curr int) int {return sum + curr}, 0)(factors) == number * 2
}

File: assign1/java/src/main/java/namcharoen_tran/PerfectNumber.java
package namcharoen_tran;

import java.util.stream.IntStream;

public interface PerfectNumber {
  public static boolean isPerfectNumberImperative(int number) {
    var sumOfFactors = 0;

    for (var i = 1; i < number + 1; i++) {
      if (number % i == 0) {
        sumOfFactors += i;
      }
    }

    return number > 1 && sumOfFactors == number * 2;
  }

  public static boolean isPerfectNumberFunctional(int number) {
    return number > 1 &&
      IntStream.rangeClosed(1, number) 
        .filter(i -> number % i == 0)
        .sum() == number * 2;
  }
}
